### ISSUE TITLE
Lack of implementation "Application flow" swagger 3.x

### ISSUE THREAD

**Issue by misi** [swagger-api/swagger-ui#2838]
The authorize button does not appear when OAuth2  Application flow is used.
I have checked the source and AFAIU it is not yet implemented..

src/core/components/auth/oauth2.jsx
flow !== APPLICATION

Please implement this feature, or help me to implement a post call inside oauth2-authorize.js
Thanks..

**Comment by ponelat** [comment]
Early on we decided against implementing the Application flow as it can only be used by mobile devices (  my understanding of oAuth 2.0 isn't perfect. so feel free to correct me ). 

As to adding a hook inside the oaut2-authorize.js, Its possible to make it a plugin. While we're working on such documentation ( plugins, etc ). You can fork, and see if you can get a hack working in the fashion that you want it. If that is successful, we can make/advise on how to make this plugin.

Happy hunting!

**Comment by misi** [comment]
Thanks for your answer!

In my understanding Application flow client_credentials is about Machine2Machine
https://oauth2.thephpleague.com/authorization-server/client-credentials-grant/

I feel it is important to have it.
The client_id and client_secret and scope need to be POST-ed to access_token AS url.

I definitely want to work on it.

Could you help little bit and point me to any example code what I can reuse for POST data ?

May adding a notice to known issues in 3.x would help for other peoples.

Thanks again for confirming the issue! 
If you could help me with few hints would be highly appreciated!

**Comment by misi** [comment]
I am wondering if sending out POST in a XMLHTTPREQUEST could fit?
like it is in the example:
https://developer.mozilla.org/en-US/docs/Learn/HTML/Forms/Sending_forms_through_JavaScript
So getting back in body the Access Token in case auth worked successfully.

**Comment by misi** [comment]
Thx @ponelat. 
I hacked using "fetch" as you proposed. 
I have now the access_token value in ouath2-authorize.js
See [my changes](https://github.com/misi/swagger-ui/commit/ac8ac340aff281445cc4dacfa03cf902c4e197ea).

**Comment by bodnia** [comment]
@misi @ponelat I believe this flow is called `Client Credentials` like here http://oauthbible.com/#oauth-2-two-legged and from this he have only `password` and `implicit` implemented. The cause of this missing feature is that in [OpenAPI](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#fixed-fields-15 ) specification there are only 4 flows allowed:
>The flow used by the OAuth2 security scheme. Valid values are "implicit", "password", "application" or "accessCode".

So in case of adding new flow it also should be added to specification.
@webron do you have any thoughts about this?

**Comment by ponelat** [comment]
So there is history here... see https://github.com/OAI/OpenAPI-Specification/issues/400 if you feel like reading.

### tl;dr

 OAS flow  | oAuth 2.0 grantType |
| --- | --------------------------- |
| implicit | Implicit |
| password | Resource Owner Password Credentials |
| application | Client Credentials |
| accessCode | Authorization Code |
  
So in short... we don't have support for `application` flow... it doesn't mean what I thought it meant ( Mobile URIs, like `facebook://` ) and I'm not sure if there is planned support. If your hack works ( I'm guessing it does... ) then this just needs to be prettied up and can likely end up in core code vs a plugin. I still don't fully understand `application` flow, but the `Client Credentials` seem straightforward.

@misi thanks for following along. I can't comment on the whether this will get in or not, but at least this discussion has been reopened :smile: 

@webron thoughts on triage? Has any work been done on oAuth in 3.0 that would make this moot?

**Comment by misi** [comment]
I updated my git fork.
I will add PR please review it. 
It works for me.
Any review comment will be highly appreciated..

**Comment by webron** [comment]
To summarize things - all 4 flows that are defined in the OAuth2 spec are supported in 2.0. The only issue is the naming convention, and as @ponelat mentioned, there's a proper mapping there.

In 3.0 we support the same 4 flows, but we renamed the flows to follow the OAuth2 spec more closely and avoid confusion.

What @misi is trying to do should ultimately be supported in Swagger-UI (and JS of course).

**Comment by misi** [comment]
@webron Many thanks for the clarifications!

I updated the PR and fixed the open issues.
It seems it doesn't have any issue according travis-ci.
I also tested it and works as expected.

Please review it...
THX!

**Comment by bodnia** [comment]
Changes merged https://github.com/swagger-api/swagger-ui/pull/2864
